### ISSUE TITLE
Fix S2386: Rule should not report when field is readonly and ...

### ISSUE THREAD

**Issue by Evangelink** [SonarSource/sonar-csharp#505]
... initialized with immutable type in ctor

### Description

When a `public static readonly` mutable field is initialized through the constructor with an immutable type the rule should not raise an issue.

### Expected behavior

Issue should not be reported.

### Actual behavior

Issue is reported.

### Related information
[RSPEC-2386](https://jira.sonarsource.com/browse/RSPEC-2386)

Related bug: #1491 (same issue, the rules share a base class)

**Comment by duncanp-sonar** [comment]
Currently the rule only looks at ```SyntaxKind.FieldDeclaration``` nodes. To fix this FP (and the one in the related bug) we would need to also check the constructor to see if the field is initialised there i.e. assigned to directly, or set as an ```out``` parameter. This would involve re-writing the rule to look at class/struct declarations, then walk the syntax tree looking for the field declarations and the constructor.
